src/extract/resolve/readPackageDeps.js

```diff
@@ -46,3 +47,5 @@ module.exports =
         }
         return lRetval;
     };
+
+module.exports = _memoize(readPackageDeps);
```

Comment:
I guess this can be optimized further. Currently this only momoizes the result for each start directory. It doesn't cache the lookups further up the directory tree.
Consider the following structure:

```
package.json
dir-a
    |-- subdir-one
    |    |-- file-one.js
    |    \-- file-two.js
    \-- subdir-two
         \-- file-three.js
dir-b
    \-- file-four.js
```

The cache will be used once because `file-one.js` and `file-two.js` are in the same directory.
That means there are still 2 lookups for `dir-a` and 3 lookups for the current directory.

If `readPackageDeps` was recursively calling a memoized version of itself, you would read only once in each directory.

```js
const readPackageDeps =
    (pFileDir) => {
        let lRetval = null;

        try {
            // find the closest package.json from pFileDir
            const lPackageContent = fs.readFileSync(path.join(lLookupDir, 'package.json'), 'utf8');
            try {
                lRetVal = JSON.parse(lPackageContent);
            } catch (e) {
                // left empty on purpose
            }
        } catch (e) {
            const lNextDir = path.dirname(lLookupDir);

            if (lNextDir !== pFileDir) {
                // not yet reached root directory
                lRetval = momoizedReadPackageDeps(lNextDir);
            }
        }
        return lRetval;
    };

const memoizedReadPackageDeps = _memoize(readPackageDeps);

module.exports = memoizedReadPackageDeps;
```

Reply:
I can't really test this, because I only have SSDs and the performance difference between completely uncached and fully cached (code I posted above) is around 100ms from 2.35s to 2.25s.
The difference between your cache implementation vs mine is not really measurable.

Reply:
- this recursive implementation is more elegant than the `do ... while` - I like it! keeper :-)
- performance-wise it indeed seems a tie between my crude memoize-only and this one (even on humongous code bases) - which is a bit puzzling.
- your measurement of cached being ~4% faster than non-cached is consistent with my findings.

The relative differences between cached and uncached on an old MacBook with spinning disks were even smaller. I might do some more tests with my current Mac & external spinning disks to level out the differences in CPU, memory, ..., but not in the context of this PR - the differences are too small to worry about for now.